File: db4on/core/src/core/com/db4o/internal/convert/Converter.cs
namespace com.db4o.@internal.convert
{
	/// <exclude></exclude>
	public class Converter
	{
		public const int VERSION = com.db4o.@internal.convert.conversions.FieldIndexesToBTrees_5_7
			.VERSION;

		private static com.db4o.@internal.convert.Converter _converter;

		private com.db4o.foundation.Hashtable4 _conversions;

		private Converter()
		{
			_conversions = new com.db4o.foundation.Hashtable4();
			com.db4o.@internal.convert.conversions.CommonConversions.Register(this);
		}

		public static bool Convert(com.db4o.@internal.convert.ConversionStage stage)
		{
			if (!NeedsConversion(stage.SystemData()))
			{
				return false;
			}
			if (_converter == null)
			{
				_converter = new com.db4o.@internal.convert.Converter();
			}
			return _converter.RunConversions(stage);
		}

		private static bool NeedsConversion(com.db4o.@internal.SystemData systemData)
		{
			return systemData.ConverterVersion() < VERSION;
		}

		public virtual void Register(int idx, com.db4o.@internal.convert.Conversion conversion
			)
		{
			if (_conversions.Get(idx) != null)
			{
				throw new System.InvalidOperationException();
			}
			_conversions.Put(idx, conversion);
		}

		public virtual bool RunConversions(com.db4o.@internal.convert.ConversionStage stage
			)
		{
			com.db4o.@internal.SystemData systemData = stage.SystemData();
			if (!NeedsConversion(systemData))
			{
				return false;
			}
			for (int i = systemData.ConverterVersion(); i <= VERSION; i++)
			{
				com.db4o.@internal.convert.Conversion conversion = (com.db4o.@internal.convert.Conversion
					)_conversions.Get(i);
				if (conversion != null)
				{
					stage.Accept(conversion);
				}
			}
			return true;
		}
	}
}

File: db4oj/core/src/com/db4o/internal/convert/Converter.java
/* Copyright (C) 2004 - 2006  Versant Inc.  http://www.db4o.com */

package com.db4o.internal.convert;

import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.convert.conversions.*;

/**
 * @exclude
 */
public class Converter {
    
    public static final int VERSION = ReindexNetDateTime_7_8.VERSION;
    
    private static Converter _converter;
    
    private Hashtable4 _conversions;
    
    private Converter() {
        _conversions = new Hashtable4();
        
        // TODO: There probably will be Java and .NET conversions
        //       Create Platform4.registerConversions() method ann
        //       call from here when needed.
        CommonConversions.register(this);
    }

    public static boolean convert(ConversionStage stage) {
    	if(!needsConversion(stage.systemData())) {
    		return false;
    	}
    	if(_converter == null){
    		_converter = new Converter();
    	}
    	return _converter.runConversions(stage);
    }

    private static boolean needsConversion(SystemData systemData) {
        return systemData.converterVersion() < VERSION;
    }

    public void register(int idx, Conversion conversion) {
        if(_conversions.get(idx) != null){
            throw new IllegalStateException();
        }
        _conversions.put(idx, conversion);
    }
    
    public boolean runConversions(ConversionStage stage) {
        SystemData systemData = stage.systemData();
        if(!needsConversion(systemData)){
            return false;
        }
        for (int i = systemData.converterVersion(); i <= VERSION; i++) {
            Conversion conversion = (Conversion)_conversions.get(i);
            if(conversion != null){
                stage.accept(conversion);
            }
        }
        return true;
    }
    
}
